File: LeetCode/Study Plan/Practice Problems/Medium/Perfect Squares/Go/perfect_squares.go
package main

import (
	"fmt"
	"math"
)

//---------------------recursive(TLE)---------------------------
func perfect_squares(n int) int {
	mp := make(map[int]int)
	sq := make([]int, int(math.Sqrt(float64(n))+1))
	for i := 1; i <= int(math.Sqrt(float64(n))); i++ {
		mp[i*i] = i * i
		sq[i] = i * i
	}
	return min_num_sq(mp, sq, n)
}

func min_num_sq(mp map[int]int, sq []int, n int) int {
	if mp[n] != 0 {
		return 1
	}
	min_num := math.MaxInt64
	for i := 1; i < len(sq); i++ {
		if n < sq[i] {
			break
		}
		val := min_num_sq(mp, sq, n-sq[i]) + 1
		min_num = min(min_num, val)
	}
	return min_num
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

//---------------------recursive(TLE)---------------------------

//---------------------memo---------------------------
func perfect_squares_memo(n int) int {
	memo := make(map[int]int)
	mp := make(map[int]int)
	sq := make([]int, int(math.Sqrt(float64(n))+1))
	for i := 1; i <= int(math.Sqrt(float64(n))); i++ {
		mp[i*i] = i * i
		sq[i] = i * i
	}
	return min_num_memo(&memo, sq, mp, n)
}

func min_num_memo(memo *map[int]int, sq []int, mp map[int]int, n int) int {
	if (*memo)[n] != 0 {
		return (*memo)[n]
	}
	if mp[n] != 0 {
		return 1
	}
	min_cnt := math.MaxInt64
	for i := 1; i < len(sq); i++ {
		if n < sq[i] {
			break
		}
		val := min_num_memo(memo, sq, mp, n-sq[i]) + 1
		min_cnt = min(min_cnt, val)
		(*memo)[n] = min_cnt
	}
	return min_cnt
}

//---------------------memo---------------------------

//------------------------a better approach-----------------------
func perfect_squares_eff(n int) int {
	dp := make([]int, n+1)
	for i := 1; i <= n; i++ {
		dp[i] = math.MaxInt64
	}
	max_ind := int(math.Sqrt(float64(n))) + 1
	sq := make([]int, max_ind)
	for i := 1; i < max_ind; i++ {
		sq[i] = i * i
	}
	for i := 1; i <= n; i++ {
		for j := 1; j < max_ind; j++ {
			if i < sq[j] {
				break
			}
			dp[i] = min(dp[i], dp[i-sq[j]]+1)
		}
	}
	return dp[n]
}

//------------------------a better approach-----------------------

func main() {
	fmt.Println(perfect_squares_eff(12))
}

File: LeetCode/Study Plan/Practice Problems/Medium/Perfect Squares/Cpp/perfect_squares.cc
#include <iostream>
#include <vector>
#include <map>
#include <cmath>

typedef std::vector<int> vi;
typedef std::map<int, int> mi;
typedef std::map<int, bool> mb;

//------------------memo still TLE-------------------------
int min_cnt(vi& sq, mb& m, mi& visited, int n) {
    if (m.find(n) != m.end()) return 1;
    if (visited.find(n) != visited.end()) return visited[n];
    int min_val = INT_MAX;
    for (int i=1; i<sq.size(); i++) {
        if (n < sq[i]) break;
        int val = 1 + min_cnt(sq, m, visited, n-sq[i]);
        min_val = std::min(min_val, val);
        visited[n] = min_val;
    }
    return min_val;
}

int perfect_squares(int n) {
    vi sq(sqrt(n)+1);
    mb m;
    mi visited;
    for (int i=1; i<=sqrt(n); i++) {
        sq[i] = i*i;
        m[i*i] = true;
    }
    return min_cnt(sq, m, visited, n);
}
//------------------memo still TLE-------------------------

//------------------efficient approach-------------------------
int perfect_squares_eff(int n) {
    vi dp(n+1, INT_MAX);
    dp[0] = 0;
    int max_ind = (int)sqrt(n) + 1;
    vi sq(max_ind);
    for (int i=1; i<max_ind; i++) sq[i] = i*i;
    for (int i=1; i<=n; i++) {
        for (int j=1; j<max_ind; j++) {
            if (i<sq[j]) break;
            dp[i] = std::min(dp[i], dp[i-sq[j]]+1);
        }
    }
    return dp[n];
}
//------------------efficient approach-------------------------

int main () {
    std::cout<<perfect_squares_eff(12)<<"\n";
}
